File: core/src/test/java/org/infinispan/test/data/Key.java
package org.infinispan.test.data;

import java.io.Externalizable;
import java.io.IOException;
import java.io.ObjectInput;
import java.io.ObjectOutput;
import java.util.Objects;

import org.infinispan.protostream.annotations.ProtoField;

public class Key implements Externalizable {
   private static final long serialVersionUID = 4745232904453872125L;

   @ProtoField(number = 1)
   String value;

   public Key() {}

   public Key(String value) {
      this.value = value;
   }

   @Override
   public boolean equals(Object o) {
      if (this == o) return true;
      if (o == null || getClass() != o.getClass()) return false;

      return Objects.equals(value, ((Key) o).value);
   }

   @Override
   public int hashCode() {
      return value != null ? value.hashCode() : 0;
   }

   public void writeExternal(ObjectOutput out) throws IOException {
      out.writeObject(value);
   }

   public void readExternal(ObjectInput in) throws IOException, ClassNotFoundException {
      value = (String) in.readObject();
   }
}

File: server/hotrod/src/main/scala/org/infinispan/server/hotrod/Key.scala
package org.infinispan.server.hotrod

import java.util.Arrays
import java.io.{ObjectOutput, ObjectInput, Externalizable}

/**
 * // TODO: Document this
 * @author Galder Zamarreño
 * @since
 */
// TODO: Make it an Externalizer once submodules can extend the marshalling framework
final class Key(var k: Array[Byte]) extends Externalizable {

   override def equals(obj: Any) = {
      obj match {
         case k: Key => Arrays.equals(k.k, this.k)
         case _ => false
      }
   }

   override def hashCode: Int = 41 + Arrays.hashCode(k)

   override def toString = {
      new StringBuilder().append("Key").append("{")
         .append("k=").append(k)
         .append("}").toString
   }

   override def readExternal(in: ObjectInput) {
      k = new Array[Byte](in.read())
      in.read(k)
   }

   override def writeExternal(out: ObjectOutput) {
      out.write(k.length)
      out.write(k)
   }
}